src/services/DomainModel.ts

```diff
@@ -0,0 +1,48 @@
+// Do not assume you can reach every node, sensor, or reading starting from a
+// gateway you're looking at. Be mindful that a Query is probably in effect
+// which has limited the extent to which we have populated all the objects into
+// memory. For example, if you're working with recent data from the  'Black
+// Pearl' node you might have info about the Gateway, but just the one
+// node, and a few days of readings. You'll find `Unknown`s at the points in the
+// tree that we don't care about in the context of an active Query.
+
+export interface Unknown {
+  // Unknown means you've reached an unexplored part of the tree, not a dead end.
+  explored: false;
+}
+
+export interface Query {
+  params: QueryParam[]; // TBD what a QueryParam is
+  results: Gateway[] | Node[] | Sensor[] | Reading[];
```

Comment:
As with the comment above, let's make it a collection. We can either have `Set<Gateway> | Set<Node>` etc... if the contents are homogenous, or we can define `QueryItem = Gateway | Node | ...` and have `Set<QueryItem>`. 

In fact I would go so far to say that `Query` itself should be generic on the type of result. That means we don't loose type safety for queries that are known to return a specific type.